blockindex/sgd.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2022 IoTeX Foundation
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+	"github.com/iotexproject/iotex-core/blockindex/indexpb"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/db/batch"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	_sgdBucket                = "sgd"
+	_sgdToHeightNS            = "hh"
+	_sgdActCreate  sgdActType = iota
+	_sgdActCall
+)
+
+var _sgdCurrentHeight = []byte("currentHeight")
+
+type (
+	// SGDIndexer is the interface for Sharing of Gas-fee with DApps indexer
+	SGDIndexer interface {
+		blockdao.BlockIndexer
+		// CheckContract
+		CheckContract(contract string) (address.Address, uint64, bool)
+		// GetSGDIndex
+		GetSGDIndex(contract string) (*indexpb.SGDIndex, error)
```

Comment:
does not need to expose this func as part of interface?

Reply:
expose this func  to get all the state of the contract, `checkContract` can only get receiver and percentage